pandas/_libs/algos.pyx

```diff
@@ -913,9 +891,11 @@ def rank_1d(rank_t[:] in_arr, ties_method='average',
                     for j in range(i - dups + 1, i + 1):
                         ranks[argsorted[j]] = i + 1
                 elif tiebreak == TIEBREAK_FIRST:
-                    if rank_t is object:
-                        raise ValueError('first not supported for '
-                                         'non-numeric data')
+                    if False:#rank_t is object:
+                        pass
+                        # FIXME: this breaks with the conditional nogil
+                        #raise ValueError('first not supported for '
+                        #                 'non-numeric data')
```

Comment:
@scoder the "kicking the tires" i mentioned in the OP largely is about this spot.  When the commented-out code is enabled here, I get a compile-time error:

```
                    if rank_t is object:
                        raise ValueError('first not supported for '
                       ^
------------------------------------------------------------

pandas/_libs/algos.pyx:895:24: Trying to acquire the GIL while it is already held.
```

Is the issue that the compiler doesn't know its in a not-nogil case if it gets here?

Reply:
I think the GIL acquisition around the `raise` might happen too early here, before the fused type specialisation, so that it always tries to acquire the GIL even if the `with nogil` was removed for the specific type.

I think Cython should mark the automatically injected `GilStatNode` as something internal in `InjectGilHandling` (module `ParseTreeTransforms.py`), and then discard the redundant ones in `GilCheck` instead of throwing this error for them. That would avoid the need for any special casing of fused functions later on. PR welcome.